File: src/queue.rs
use std::cell::RefCell;
use std::rc::Rc;

type Link<T> = Option<Rc<RefCell<QNode<T>>>>;

#[derive(Debug)]
pub struct QNode<T> {
    value: T,
    next: Link<T>,
}

impl<T> QNode<T> {
    pub fn new(item: T) -> Rc<RefCell<QNode<T>>> {
        Rc::new(RefCell::new(QNode {
            value: item,
            next: None,
        }))
    }
}

#[derive(Debug)]
pub struct Queue<T> {
    pub length: i32,
    head: Link<T>,
    tail: Link<T>,
}

impl<T: Copy> Queue<T> {
    pub fn new() -> Queue<T> {
        Queue {
            length: 0,
            head: None,
            tail: None,
        }
    }
    pub fn enqueue(&mut self, item: T) {
        let node = Some(QNode::new(item));
        match self.tail.take() {
            Some(current) => {
                current.borrow_mut().next = node.clone();
                self.tail = node;
                self.length += 1;
            }
            None => {
                self.head = node.clone();
                self.tail = node;
                self.length += 1;
            }
        }
    }
    pub fn dequeue(&mut self) {
        match self.head.take() {
            Some(head) => {
                self.head = Rc::try_unwrap(head)
                    .ok()
                    .expect("unable to dequeue")
                    .into_inner()
                    .next;
                self.length -= 1;
            }
            _ => {}
        }
    }
    pub fn peek(mut self) -> Option<T> {
        if let Some(v) = self.head.take() {
            Some(
                Rc::try_unwrap(v)
                    .ok()
                    .expect("something went wrong")
                    .into_inner()
                    .value,
            )
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn create_simple_queue() {
        let mut sq: Queue<i32> = Queue::new();
        sq.enqueue(32);
        sq.enqueue(13);
        sq.enqueue(3);
        sq.enqueue(330);
        dbg!(sq);
    }
    #[test]
    fn peek_simple_queue() {
        let mut sq: Queue<i32> = Queue::new();
        sq.enqueue(32);
        sq.enqueue(13);
        sq.enqueue(3);
        sq.enqueue(330);
        let res = sq.peek();
        dbg!(res);
    }
    #[test]
    fn dequeue_simple_queue() {
        let mut sq: Queue<i32> = Queue::new();
        sq.enqueue(32);
        sq.enqueue(13);
        sq.enqueue(3);
        sq.enqueue(330);
        sq.dequeue();
        dbg!(sq);
    }
}

File: structures/queue.go
package structures

// Node - Node is a data type that is contained by the Queue
type Node struct {
	data int
	next *Node
}

func newNode(data int) *Node {
	return &Node{
		data: data,
		next: nil,
	}
}

// Queue - Queue is a common data structure that is used in programming based on a linked list
type Queue struct {
	head   *Node
	tail   *Node
	length int
}

func newQueue() Queue {
	return Queue{
		head:   nil,
		tail:   nil,
		length: 0,
	}
}

func (q *Queue) enqueue(data int) {
	node := newNode(data)
	if q.tail != nil {
		q.tail.next = node
		q.tail = node
		q.length++
	} else {
		q.head = node
		q.tail = node
		q.length++
	}
}

func (q *Queue) dequeue() int {
	if q.head != nil {
		data := q.head.data
		q.head = q.head.next
		q.length--
		return data
	}

	return 0
}

func (q *Queue) peek() int {
	if q.head != nil {
		return q.head.data
	}

	return 0
}
